test/cpp/api/misc.cpp

```diff
@@ -83,7 +83,47 @@ TEST_F(AutogradTest, CanPassCustomGradientInputs) {
   ASSERT_TRUE(x.grad().allclose(y * 2));
 }
 
+// This wrapper can be instantiated at the beginning of a test to
+// capture the current deterministic flag value. When it is
+// destructed, it restores the deterministic flag value that
+// it captured
+class DeterministicTestWrapper {
+public:
+  DeterministicTestWrapper() :
+      context(&at::globalContext()),
+      deterministic_restore(context->deterministic()) {
+    // Need to set an environment variable if CUDA >= 10.2 so that we
+    // don't get an error from Context::alertCuBLASConfigNotDeterministic
+    if (context->hasCUDART() && (context->versionCUDART() >= 10020)) {
+      // Note: It would be desirable to reset this variable in the
+      // destructor, but doing so has some annoying challenges.
+      // One of which is that putenv requires that the buffer given
+      // to it remain allocated and unchanged until putenv is called
+      // again. It is possible to accomplish that, but at the moment,
+      // we have no real reason to reset the variable. That could
+      // change though.
+#if (defined(_WIN32) || defined(_WIN64))
+      _putenv(const_cast<char*>(cublas_config_override));
+#else
+      putenv(const_cast<char*>(cublas_config_override));
```

Comment:
I'm a bit dubious about this. (1) It's pretty weird and unexpected for code inside a C++ executable to be modifying the process environment, (2) in many situations, this doesn't even do anything, if the relevant code in the library reads the environment once at say startup time (this is what happens with CUDA), and (3) since you don't restore it at the end of the test, it could affect subsequent tests numerics in hard to understand and debug ways. Can we just not do this?

Reply:
Yeah that's true. The only reason I was setting this variable was to avoid the cuBLAS related error when calling `at::setDeterministic()` in the unit tests. But after changing that error to be more lazy, setting the variable here is unnecessary.